### ISSUE TITLE
Delayed execution/hijacking vulnerability

### ISSUE THREAD

**Issue by nns2009** [ton-blockchain/multisig-contract-v2#47]
# High-level overview
1. Attacker controls N-out-of-M Signers or can trick sufficient number of them to sign a specific Order (later: Exploit Order)
2. Attacker executes the Exploit Order
3. Attacker transfers control (e.g. sells) of some (but not all) of Signers

**Expected behaviour**: in order to get any Order executed, sufficient number of newly appointed Signers must sign.
**Reality**: given a specific [Exploit] Order from step #2 was executed, it’s enough for only a single of newly appointed Signers to sign to get the pre-crafted [Malicious] Order executed.

# Specific money-grab scenario 

Attacker (or his friends) is selling signing positions in Multisig. This can be done in any way: in person in a cafe or through a secure smart contract, which temporarily assumes control (becomes a single signer) of the Multisig and subsequently transfers control, when payment is received - similar to NFT sales.
Regardless, once payment is received, the Attacker finishes execution of the [Malicious] Order prepared at the step #2 above, which can do whatever: reassign Signers, withdraw all the balance or the assets, which are under Multisig’s control.

# How it works
## Multisig can sign Orders on behalf of anyone

Given N-out-of-M Signers sign an order, Multisig can send **any** messages (later: Malicious Messages), which can include messages to Order smart contracts and most importantly: **can sign an Order on behalf of any Signer**. That is because of the `op::init` logic in the Order contract - the only authentication in this operation is that the message is from the corresponding Multisig:

    if (op == op::init) {
      throw_unless(error::unauthorized_init, equal_slices_bits(sender_address, multisig_address));

Then, in case of Order’s first init:

    int approve_on_init? = in_msg_body~load_bool();
    if (approve_on_init?) {
        int signer_index = in_msg_body~load_index();
        add_approval(signer_index);
        try_execute(query_id);
    }

which means that a message from Multisig can pretend to be from any Signer by specifying the necessary index. The same in case of Order’s further inits:

    int approve_on_init? = in_msg_body~load_bool();
    throw_unless(error::already_inited, approve_on_init?);
    int signer_index = in_msg_body~load_index();
## Multisig can print any Orders out of thin air

Furthermore, these Malicious Messages from Multisig are not limited by creating Orders with its real current parameters: they can request to create **any** Order, with any `threshold`, `signers`, `expiration_date` and `order`.

## Attack preparation
- The Attacker prepares an Exploit Order
- gets N-out-of-M Signers to sign it, so it gets executed
- Multisig sends Malicious Messages
- which create Malicious Orders:
    - with `signers` dict containing the addresses to which the control is expected to be transferred
    - `threshold` >= `N_new`, where N_new is expected new threshold after the transfer of control, or `threshold` can just be set as `M_new` to be sure
    - `expiration_date` - long enough
    - `order` - whatever the Attacker wants: can transfer the control of Multisig back or withdraw balance or mess with its assets
- and sends other Malicious Messages, which “sign” (approve) each of these Orders on behalf of `threshold-1` Signers (but leaving at least one “signing position” controlled by the Attacker unsigned)

I say Order**S** (in multiple) because the exact new set of “Signers” might not be known. In case of selling through a secure smart contract and subsequent immediate hijack: the next (after payment) Multisig config might be known exactly, but in cases of:

- Multiple potential buyers
- Other Signers partially administering the deal
- Delayed hijack

the exact set of Signers might not be known and the Attacker would need to cover all the most probable options.
**For example**
Multisig is currently controlled by Alice, Bob and Clara. Alice is the Attacker.
Clara is expected to transfer control to either Dan or Jack (later X=”Dan or Jack”).
It is also not known in which order Alice, Bob and X would be arranged in the new Signers dictionary.
So the Attacker constructs 12 Malicious Orders (2 options for new Signer * 6 permutations) each with a threshold of 3, for various possible set of Signers:

- Alice, Bob, Dan
- Alice, Dan, Bob
- …
- Alice, Bob, Jack
- Alice, Jack, Bob
- …

and makes an Exploit Order, which “signs” each of these Malicious Orders with 2 others’ approvals (Bob and X).

## Attack (some time later)

Control is transferred, money is received or whatever the Attacker was waiting for happened.
The Attacker signs (legitimately, more or less) one of the pre-crafted Malicious Orders (the one with the matching set of Signers), which causes it to reach its `threshold` and send execution request to the Multisig.
Attacker previously chose large (enough) `expiration_date` and `threshold` when creating Malicious Orders, so all the checks on Multisig pass and the corresponding Malicious Order gets executed.

## After the attack

The Attacker celebrates 😈 🤑 💰 💵 😎
TON community is devastated 😕 🙁 ☹️ 😣 😖 😫 😩 🥺 😢 😭

# Potential fixes

Having super clear UI about what you sign is not enough as the setup described might also be achieved with N Signers participating knowingly and voluntarily.
Even with super clear UI, social engineering of current Signers should never impact future Signers.

Controlling what messages come out of the Multisig doesn’t seem practical (or at least not immediately obvious how to implement it).

The most promising approach seems to be to tighten the 2nd Guarantee:

> Change of the signers set invalidates all orders with other set. More strictly, Order is only valid when current signers are equal to signers at the time Order was created.

and make it so that **any** change of Signers invalidates **all** Orders.

One way to achieve this is: is to introduce “Signers/Config last update timestamp”, which is kept on Multisig and “Order creation timestamp” and have Multisig reject any execution requests in case the received Timestamp from Order is < than its own timestamp. Importantly, the Order should initialize this Timestamp on its own (e.g. with `now()`), instead of receiving it from the Multisig (in which case, it can also be faked inside Malicious Messages).

**Comment by nns2009** [comment]
Alternative (mostly) fix: not allow "auto signing" coming from Multisig. Only accept direct messages from Signers themselves (thus the one who made a proposal, would need to send a separate message to sign it).

**Comment by drinkius** [comment]
For malicious order to be executed in the multisig - current multisig owners list & their order must match the list of owners in the order exactly, since upon execution the hashes of the cells containing lists are compared:
https://github.com/ton-blockchain/multisig-contract-v2/blob/107ee13aa4cbabdc9ff0684b738dcd272c4211bc/contracts/multisig.func#L166
In general this scenario revolves around people not understanding what they are signing - so it's just social engineering, not exploit of the contract's logics, if I understood everything correctly. You'd still need to gather all the signatures from others in order to perform malicious action

**Comment by nns2009** [comment]
> For malicious order to be executed in the multisig - current multisig owners list & their order must match the list of owners in the order exactly

Yes - and the attacker (under some realistic circumstances) can make it happen, as I described.

> In general this scenario revolves around people not understanding what they are signing ...

No - everyone malicious can participate willingly and voluntarily.
Furthermore, malicious participants (by examining the "Exploit Order" they sign) can know exactly what kind of orders the Attacker will be able to execute, so they actually have some good guarantees about what the Attacker can or can't do. Thus they can conspire with the Attacker (e.g. be bribed) without the need to trust him. For example: they can sign the "Exploit Order" that would allow the Attacker to re-assign Signers to the old (pre-sale) ones and be absolutely sure that they would keep a position as a Signer after the hijacking.
Even more: all of them can later claim that they "have been tricked, we didn't know what we were signing" and that they supposedly "didn't conspire".

> You'd still need to gather all the signatures from others in order to perform malicious action

The Attacker would need to gather enough signatures to execute "Exploit Order".
The Attacker would **only need his own** signature to execute "Malicious Orders".

**Comment by nns2009** [comment]
# In other words
The vulnerability I described breaks two important things about Multisig logic:

1) The 2nd Guarantee:

> Change of the signers set invalidates all orders with other set. More strictly, Order is only valid when current signers are equal to signers at the time Order was created.

The "Malicious Orders" are not going to be invalidated despite the fact that the current Signers are different from the Signers at the time of Order creation, because "Malicious Orders" are deployed to have future expected set of Signers (and not the actual one at the moment of deployment).

2) The very fundamental behind the Multisig intention:

> at least N parties out of predefined set of M signers must approve Order to execute it

"Malicious Orders" can have their `N-1` signatures pre-"signed" (even on behalf of parties that are not part of the conspiracy/attackers) so only a single genuine signature is necessary.

**Comment by drinkius** [comment]
> The "Malicious Orders" are not going to be invalidated despite the fact that the current Signers are different from the Signers at the time of Order creation, because "Malicious Orders" are deployed to have future expected set of Signers (and not the actual one at the moment of deployment).

Actually not action can be performed with the order or it's execution if at the time of signing the signers set in multisig is different from the set of signers in the order (including the positions of each signer) since the cells are compared by hashes. So I don't really see where this security guarantee is broken

I think the main issue lies here:

```
Multisig sends Malicious Messages
which create Malicious Orders:
with signers dict containing the addresses to which the control is expected to be transferred
```

Here you assume that orders with mismatched list of signers can be created, but in reality - multisig always creates orders only with the current list of signers it holds. So no, orders with future sets of signers can't actually be created per current implementation, at least I couldn't find was to do so.

This is the culprit:
https://github.com/ton-blockchain/multisig-contract-v2/blob/1e6f91942afb4515f4a120c2bb80825bf0eeb01c/contracts/multisig.func#L104

The list of signers that's sent to the order is being fetched from the current multisig state

**Comment by nns2009** [comment]
> multisig always creates orders only with the current list of signers it holds

No: Multisig logic only creates Orders with the current list of Signers, but the [Exploit] Orders (with a proper current list of Signers), which Multisig executes, can send arbitrary messages and thus can create [Malicious] Orders with an arbitrary set of Signers.

Furthermore, in universal case it is not possible to prevent this: since it may be necessary to send messages to some third-party smart contracts that look exactly the same as the Order creation messages and it is impossible to determine by the destination address whether it belongs to the Order.

The critical issue is that such [Malicious] Orders are not invalidated when Signers change.

**Comment by drinkius** [comment]
> it is impossible to determine by the destination address whether it belongs to the Order

Actually it is and can be easily done in the UI, plus it should be part of due diligence while signing orders with current multisig design

I'm still convinced that such vulnerabilities would fall under "social engineering" category since the main disclaimer is - people should know what they are signing. If we assume that the attacker somehow can force others to sign arbitrary orders without checking the contents - then anything is possible. The attack really stands only if it can be executed without other people's involvement

**Comment by ProgramCrafter** [comment]
I hold opinion that whether it is more related to social engineering or not, guarantee that unexpected actions won't happen **can** be given, with only a slight change in architecture.

I model people's wallet software as showing amount of assets which would be sent, changes to validator set, but not displaying what contracts would be deployed if any. Doing otherwise would hinder mass adoption.

**Comment by drinkius** [comment]
Here I agree, it's better to bring the risks to a minimum

**Comment by nns2009** [comment]
> Actually it is and can be easily done in the UI ...

How? Give us the algorithm.

> it should be part of due diligence while signing orders

But regardless of the UI-algorithm, hackers don't need UI. They would totally do their due diligence and send arbitrary messages all they want, ignoring the UI checks. So here (from my previous message): "it is impossible to determine by the destination address whether it belongs to the Order" - I was talking about the check on the smart contract - the one, which hackers can't ignore.

> I'm still convinced that such vulnerabilities would fall under "social engineering" category

Please re-read my messages in this thread because I wrote everything in detail (starting with "High-level overview", "Specific money-grab scenario", which outline scenarios that shouldn't be possible) and already [answered](https://github.com/ton-blockchain/multisig-contract-v2/issues/47#issuecomment-2048293216) you about social engineering.

> since the main disclaimer is - people should know what they are signing ...

No, the main problem is that current Signers can "sign" (and subsequently execute) Orders on behalf of future Signers.

> It's quite easy to create payloads that when executed - transfer money to the attacker, obfuscated or not

Yes, for this specific reason:

- It's ok that the current Signers can hack themselves as much as they want
- It is absolutely **not** ok that the current Signers can hack the future Signers

Based on your message, it seems that you don't understand the nature of the attack, so I'll try to rephrase the most basic scenario:

1. I control at least N-of-M Signer positions of an important Multisig
2. I (the Attacker) sign an [Exploit] Order (voluntarily, no trickery, no social engineering!)
3. I sell some of my N Signer positions to you through a secure smart contract, which guarantees that ones you transfer money, the Signer config will change to the one, which includes you
4. You transfer money, being confident in the Sale smart contract
5. Sale smart contract indeed changes Signer config to the ones, which includes you, so I (the Attacker) don't have necessary N Signer positions anymore in the new config
6. Nevertheless, I execute [Malicious] Order, which does whatever I want (e.g. changes the Signer config to the old, pre-sale one) with only a single (my) signature, without the approval of N Signers
7. You are surprised, because #6 should absolutely not have been possible

No social engineering, no tricked Signers, everything voluntarily.

I hope this clears things up

**Comment by nns2009** [comment]
I actually realized [this](https://github.com/ton-blockchain/multisig-contract-v2/issues/47#issuecomment-2052445266):

> Furthermore, in universal case it is not possible to prevent this [about the Multisig being able to create Malicious Orders]

might not be true. It's still not impossible to determine by the destination address whether it belongs to the Order (although I am still curious if you can come up with such an algorithm). But any Order deployment would need to contain its state_init, which among other things, would contain Deployed-Smart-Contract's code. Multisig can reject any outgoing messages with state_init's Deployed code equal to Order's code. It seems like it could work, but I don't know for sure if some kind of trickery with library functions is (not)possible to produce the same final code starting from two distinct state_init's deployed codes.

Still not saying it's the best solution, but looks like it not as impossible as I used to think.

**Comment by tolya-yanot** [comment]
You've touched on an interesting scenario for social engineering.

We are talking about social engineering because the scenario only works for the **sale** of a multisig by attackers to new owners, with one of the attackers having to stay with the new set of owners and the new owners not having to verify the pending orders upon acceptance the deal. 

The contents of all orders are available on the blockchain and all such orders can be found from outgoing transactions from the multisig. So we are talking about a social relationship (sale) and buyer inattention. The buyer might as well fail to check that he is on the list of new owners or get unrelevant information on a phishing site.

We don't see the application in other scenarios.

If owners change the configuration for their needs - this case doesn't apply.

Although the contract logic matches the description (orders can contain arbitrary actions, orders are executed when signers and threshold match) and there is no award for social engineering in the contest rules, we would like to give you a bonus of **1000 TON** (**6600$** at the moment) for pointing this out.

**Comment by nns2009** [comment]
> You've touched on an interesting scenario for social engineering.

I am a bit shocked you also dismiss it as “social engineering”.


> because the scenario only works for the **sale** of a multisig by attackers to new owners

Not necessarily sale, might be a simple transfer, with current Signers leaving themselves a “hack” for a future. Regardless, being able to change Signers is intended.


> with one of the attackers having to stay with the new set of owners …
> If owners change the configuration for their needs - this case doesn't apply.

Yes, it is not unconditional “everyone can steal Multisig from absolutely anyone at any time” type of vulnerability, but I don’t claim it to be (and obviously don’t claim to deserve the full 20k TON prize). That said, most real-life vulnerabilities have some sort of limitations. Partial sale (keeping at least one spot) is quite a realistic scenario. So limitations here are not far-fetched.


> So we are talking about a social relationship (sale) …

“social relationship” as in “interaction of several parties” (totally normal thing to happen), not “social engineering” as in “trickery” to sign something unintended


> … and buyer inattention

Given the description in the repository, it’s not “buyer inattention”, it’s trust in the given specifications. Having to check every single pending Order from the entire history of Multisig in question is simply not necessary per description, which clearly states:

> Change of the signers set invalidates all orders with other set. More strictly, Order is only valid when current signers are equal to signers at the time Order was created.

---
> Although the contract logic matches the description …

No, it doesn’t. It matches the specific part you quoted, but it violates the 2nd Guarantee:

> Change of the signers set invalidates all orders with other set. More strictly, Order is only valid when current signers are equal to signers at the time Order was created.

---
> … and the new owners not having to verify the pending orders upon acceptance the deal.

The fact that you say such verification is necessary is admitting that 2nd Guarantee is not satisfied, according to which “the pending orders” should simply not be valid anymore after transfer.


> The contents of all orders are available on the blockchain and all such orders can be found from outgoing transactions from the multisig.

What you are talking about here is that there is a solution, which doesn’t involve any update to the smart contracts: asking Multisig-wallet apps to parse all the related parts of blockchain and show every single pending order from the entire history. But then (very importantly!) **updating the description correspondingly**: removing the 2nd Guarantee from it entirely and replacing it with a warning that **the change of Signers DOES NOT guarantee Order invalidation** and that in fact some the pending Orders might have already been signed on your behalf!

But, **as per contest description of Multisig**, it **MUST** guarantee that all pending Orders become invalid when Signers change. I showed that it doesn’t. It looks like a very clear mismatch to me.

I showed that the 2nd Guarantee is not satisfied and either:

- change of the smart contracts (to satisfy it) or
- change of the description (removal of the 2nd Guarantee with, consequently, all the necessary pre-transfer/sale verifications (you talked about) being necessary)

is necessary.
